newlog infra changes on EVE

naiming-zededa opened the pull request:
- commit for newlog infra changes
- commit for vendor files in pkg/newlog
- change LOGGING.md to modify some tuned numbers
- run newlog scheme by default, cloud side code was already in

rvs commented on images/rootfs.yml.in:
```diff
@@ -30,6 +30,9 @@ services:
      image: RSYSLOGD_TAG
      cgroupsPath: /eve/services/rsyslogd
      oomScoreAdj: -999
+   - name: newlogd
```
Why are we not removing rsyslogd?

eriknordmark replied:
@rvs I thought we have a knob to switch back to rsyslog as we soak test newlogd

naiming-zededa replied:
@rvs this is short term, most of the clusters not upgraded to the newlog code yet, also for testing now. will remove later.

rvs commented on pkg/newlog/Dockerfile:
```diff
@@ -0,0 +1,23 @@
+COPY ./  /newlog/.
+WORKDIR /newlog
+
+RUN go build -mod=vendor -o newlogd ./cmd
```
what's the additional size impact of this container on rootfs and what kind of memory footprint are you seeing @naiming-zededa ?

naiming-zededa replied:
looked at the size of xen/kvm squash images, about 260Mbytes, on my device running debug, i got top showing newlogd with 111m in VSS,

eriknordmark replied:
What is the RSS for rsyslogd in comparison? My understanding is that rsyslogd isn't started unless forced with the /config file. but running both of them might push small systems over the edge.

naiming-zededa replied:
only the docker part is running, not the daemon itself for rsyslog. below is example on ps grep for rsyslog
1262 root      0:00 /usr/bin/containerd-shim-runc-v2 -namespace services.linuxkit -id rsyslogd -address /run/containerd/containerd.sock
and the RSS is 1060:
ps -o pid,vsz,rss | grep 1262
 1262 110m 1060

rvs replied:
111m of heap is really not ideal since we're trying to stay under 500Mb @naiming-zededa -- not sure what to do about this just yet -- but it seems like a big deal to me and would definitely affect our ability to deliver on Dell 300x systems.

naiming-zededa replied:
@rvs @eriknordmark the 111m is the virtual memory the 'go' program allocate, even I made this newlogd to be only a simple for loop, it would have 100m VSZ reported. I also found if change to go version 1.15, the VSZ reported would shoot up to 700m! but the RSS, the usage of the memory is about 10m, and with the newlogd message is limited by the channel message, it will not grow much for real memory usage.

rvs reviewed (changes requested):
This is very exciting @naiming-zededa ! I'll give it a much more thorough review tomorrow, but for now just a few quick comments inline and one observation: it appears that some yetus golintci complaints are pretty legit. Do you want to take a look into that?

eriknordmark commented on pkg/pillar/scripts/device-steps.sh:
```diff
@@ -468,6 +468,13 @@ if ! pgrep logmanager >/dev/null; then
+    echo "$(date -Ins -u) Starting loguploader"
+    $BINDIR/loguploader &
+    wait_for_touch loguploader
+    touch "$WATCHDOG_FILE/loguploader.touch"
```
Are you touching loguploader.pid in some existing loop?

naiming-zededa replied:
have a "<-stillRunning.C" in the loguploader's main loop

eriknordmark commented on pkg/pillar/types/newlogtypes.go:
```diff
@@ -0,0 +1,71 @@
+	NumInputEvent     uint64 // total event input from log source, device include kmessages
+	// from loguploader
+	NumGZipFileRetry      uint64    // total gzip file upload retries, device
+	NumGZipFileDrop       uint32    // total gzip file upload 4xx failure and kept on device
```
name has "Drop" but the comment says "kept". Confusing.

naiming-zededa replied:
Yes. Trying to figure out a good term. we will not send this anymore, and it's kept only on device;-) Maybe changing this name to "NumGZipFileKeptLocal"

eriknordmark replied:
That name makes a lot more sense.

eriknordmark commented on pkg/pillar/types/newlogtypes.go:
```diff
@@ -0,0 +1,71 @@
+
+type logfileMetrics struct {
+	// from newlogd
+	NumGZipFilesSent  uint64 // total gzip files uploaded, device
+	NumGZipBytesWrite uint64 // total gzip log in bytes, device
+	NumBytesWrite     uint64 // total log bytes write to file before gzip, device
+	NumGzipFileInDir  uint32 // current number of gzip files remain, device
+	NumInputEvent     uint64 // total event input from log source, device include kmessages
```
Does "device" here mean it excludes the app logs? Or something else?

naiming-zededa replied:
Correct. 'device' means not App log related.

eriknordmark replied:
That appears to be incorrect, because you use this struct as 
	DevMetrics logfileMetrics // Device metrics
	AppMetrics logfileMetrics // App metrics

So you can drop the ", device" from the comments and maybe say before the logfileMetrics that it is used to separately track device and app metrics?

naiming-zededa replied:
makes sense, remove the 'device' in comments of the struct.

eriknordmark commented on pkg/pillar/types/newlogtypes.go:
```diff
@@ -0,0 +1,71 @@
+
+type serverStats struct {
+	CurrCPULoadPCT  float32 // newlog server CPU percentage usage
+	AvgCPULoadPCT   float32 // newlog server CPU avarage load percentage
```
typo

naiming-zededa replied:
ok.

eriknordmark commented on pkg/pillar/types/newlogtypes.go:
```diff
@@ -0,0 +1,71 @@
+type serverStats struct {
+	CurrCPULoadPCT  float32 // newlog server CPU percentage usage
+	AvgCPULoadPCT   float32 // newlog server CPU avarage load percentage
+	CurrProcessMsec uint32  // newlog server process log duration in msec
```
Does "current" mean from most recent attempt?

naiming-zededa replied:
Yes, the 'Last' one

eriknordmark commented on pkg/pillar/types/newlogtypes.go:
```diff
@@ -0,0 +1,71 @@
+	MinUploadMsec  uint32 // min upload to cloud delay in msec
+	MaxUploadMsec  uint32 // max upload to cloud delay in msec
+	AvgUploadMsec  uint32 // avg upload to cloud delay in msec
+	CurrUploadMsec uint32 // current upload to cloud delay in msec
```
Does "current" mean from most recent attempt?

naiming-zededa replied:
Correct.

eriknordmark commented on pkg/pillar/types/newlogtypes.go:
```diff
@@ -0,0 +1,71 @@
+// NewlogMetrics - Metrics from newlogd and loguploader
+type NewlogMetrics struct {
+	// logupload signal to newlogd
+	FailedToSend      bool      // loguploader failed to send to cloud
```
boolean and not a count of how many times it failed?

naiming-zededa replied:
there is a local variable keep track of how many times failed, this is for publication from loguploader to newlogd that it'
s currently in or not in 'FailedToSend' state.

eriknordmark replied:
ok

naiming-zededa commented:
> This is very exciting @naiming-zededa ! I'll give it a much more thorough review tomorrow, but for now just a few quick comments inline and one observation: it appears that some yetus golintci complaints are pretty legit. Do you want to take a look into that?

I'll work on that yetus, mostly not check return values, which I followed most usages in EVE.

naiming-zededa commented:
@rvs @eriknordmark the three yetus errors on sync.Mutex, this is because I'm reuse the api like log.LogBundle so the device and cloud side don't have to redefine them for the newlog. When the api generated the go functions and structures, it probably has the sync in them. But i'm here doing the json.Marshal() for the structure to format the string output of json, we don't use the sync here for any purpose. they can be ignored

gkodali-zededa commented on images/docker-compose.yml.in:
```diff
@@ -52,6 +52,17 @@ services:
+            - persist:/persist
+            - config:/config
+            - run:/run
+        # network_mode: host
```
can this line be removed now?

naiming-zededa replied:
Sure. will do.

eriknordmark commented on pkg/newlog/cmd/newlogd.go:
```diff
@@ -0,0 +1,1281 @@
+		log.Fatal(err)
+	}
+	tt := timestamp.Timestamp{Seconds: t.Unix(), Nanos: int32(t.Nanosecond())}
+	return tt
```
Can't you return a pointer here to avoid copying the lock?

naiming-zededa replied:
ok. let me try that to see if the yetus on this goes away

eriknordmark commented on pkg/newlog/cmd/newlogd.go:
```diff
@@ -0,0 +1,1281 @@
+				Filename:  entry.filename,
+				Msgid:     msgIDCounter,
+				Function:  entry.function,
+				Timestamp: &timeS,
```
Do you need this timestamp to be in protobuf format as opposed to a string with RFC3339? What does the receiver in the controller do with the protobuf-encoded timestamp?

eriknordmark replied:
And why use a struct generated from protoc? Can't you define a normal go struct for this? I don't know what other public fields (apart from the lock) which you will json encode from the protoc generated struct.

naiming-zededa replied:
on the controller side, basically this: time.Unix(entry.Timestamp.Seconds, int64(entry.Timestamp.Nanos)) after the json Unmarshal. again, this is to reuse all the existing APIs of logging for EVE/controller with no change.

eriknordmark commented on pkg/newlog/cmd/newlogd.go:
```diff
@@ -0,0 +1,1281 @@
+		return
+	}
+	devMetaData.uuid = status.DeviceUUID.String()
+	logmetaData = formatAndGetMeta("")
```
What happens to logmetaData? Does it have to be a global variable?

naiming-zededa replied:
here we just format the device log related meta-data (non-app, so no app-uuid string). i don't have context structure in newlogd since it's a simple one file program.

eriknordmark replied:
I don't understand from the local context what logmetaData is for. Makes it harder to understand because it is a global variable. I guess I have to grep for it.

naiming-zededa replied:
ok. i'll put some comments there. it's a string composed of device-uuid, eve-release and partition number. so the device subscription calls can update the fields, it's being used for later every new logfile to include this metadata.

eriknordmark commented on pkg/newlog/cmd/newlogd.go:
```diff
@@ -0,0 +1,1281 @@
+			}
+		}
+	}
+	metaStr := logs.LogBundle{
```
Ditto; json encode of protoc generated struct doesn't make much sense.

naiming-zededa replied:
@eriknordmark i tried to reuse the api defined for current log, just happens the json also was handled in the generated go files. otherwise, i need to define some new go api with json which i don't see anyone else in EVE/cloud is doing it

eriknordmark replied:
Shouldn't the protoc internal fields appear in the json encoded output? If not, then why is it complaining about copying the lock?

naiming-zededa replied:
they are not in the final json representation string, due to the 'omitempty' definition. but i guess they have some intermediate steps may do copy.

eriknordmark reviewed (commented):
Design question: aren't dnsmasq, dhcpcd, radvd etc sending their log messages to the syslog UDP port on localhost? Or how does newlogd pick them up?

naiming-zededa commented:
> Design question: aren't dnsmasq, dhcpcd, radvd etc sending their log messages to the syslog UDP port on localhost? Or how does newlogd pick them up?

all those i think handled from containerd which takes the stdout/stderr and converting them into memlogd input. this is the same for current log pipeline.

rvs commented:
> Design question: aren't dnsmasq, dhcpcd, radvd etc sending their log messages to the syslog UDP port on localhost? Or how does newlogd pick them up?

@eriknordmark -- at this point we don't have any freestanding daemons left -- everything gets plumbed through individual containers->containerd->memlogd

naiming-zededa commented:
I have tried this newlogd image on Dell 3003 device and running for 15 hours, the memory condition is no worse than the normal image.

rvs commented:
> I have tried this newlogd image on Dell 3003 device and running for 15 hours, the memory condition is no worse than the normal image.

Great @naiming-zededa -- thanks for checking this. In that case, let me do one more pass through this on Wed and I think it can be merged (@eriknordmark -- anything else aside from the feedback you already provided)?

rvs commented:
Hey @naiming-zededa -- looks like Eden tests are failing on this one -- this will have to be investigated

naiming-zededa commented:
> Hey @naiming-zededa -- looks like Eden tests are failing on this one -- this will have to be investigated

@rvs the ouput is below, is it possible that the eden is expect the log which will not get in this newlog? (since the api has changed)
=== RUN   TestLog
            lim_test.go:143: Wait for log of default number=1 timewait=600
            testContext.go:252: WaitForProc terminated by timeout 10m0s
        --- FAIL: TestLog (600.00s)
        FAIL
        [exit status 1]

rvs commented:
I think that's exactly the issue @naiming-zededa -- lets fix that before this can be merged.

eriknordmark commented:
> all those i think handled from containerd which takes the stdout/stderr and converting them into memlogd input. this is the same for current logging pipeline. The syslog UDP port is between rsyslogd -> logmanager, as a way for logmanager to be the 'server' of this rsyslog.

I don't think so. @gkodali-zededa don't we depend on rsyslogd listening on the syslog UDP port?
@naiming-zededa you should see some dnsmasq logs in the backend. Do you see those with newlog?

gkodali-zededa commented:
> all those i think handled from contain



> > all those i think handled from containerd which takes the stdout/stderr and converting them into memlogd input. this is the same for current logging pipeline. The syslog UDP port is between rsyslogd -> logmanager, as a way for logmanager to be the 'server' of this rsyslog.
> 
> I don't think so. @gkodali-zededa don't we depend on rsyslogd listening on the syslog UDP port?
> @naiming-zededa you should see some dnsmasq logs in the backend. Do you see those with newlog?

Yes, newlog will have to read from unix socket for logs from dnsmasq, dhcpcd etc.
```module(load="imuxsock" SysSock.Name="/dev/log")```
This is what rsyslogd does.

naiming-zededa commented:
> > all those i think handled from contain
> 
> > > all those i think handled from containerd which takes the stdout/stderr and converting them into memlogd input. this is the same for current logging pipeline. The syslog UDP port is between rsyslogd -> logmanager, as a way for logmanager to be the 'server' of this rsyslog.
> > 
> > 
> > I don't think so. @gkodali-zededa don't we depend on rsyslogd listening on the syslog UDP port?
> > @naiming-zededa you should see some dnsmasq logs in the backend. Do you see those with newlog?
> 
> Yes, newlog will have to read from unix socket for logs from dnsmasq, dhcpcd etc.
> `module(load="imuxsock" SysSock.Name="/dev/log")`
> This is what rsyslogd does.

@gkodali-zededa ok, this 'newlogd' used to have read from the /dev/log, but i monitored that and didn't see anything, I thought it was not needed. I'll put it back.

naiming-zededa commented:
> > all those i think handled from containerd which takes the stdout/stderr and converting them into memlogd input. this is the same for current logging pipeline. The syslog UDP port is between rsyslogd -> logmanager, as a way for logmanager to be the 'server' of this rsyslog.
> 
> I don't think so. @gkodali-zededa don't we depend on rsyslogd listening on the syslog UDP port?
> @naiming-zededa you should see some dnsmasq logs in the backend. Do you see those with newlog?

let me recheck on this., the rsyslogd did have a "/dev/log" in configure, I didn't see things from that channel and i removed that. I'll put it back to see then.

giggsoff commented on pkg/pillar/cmd/loguploader/loguploader.go:
```diff
@@ -0,0 +1,744 @@
+			log.Fatal(err)
+		}
+		appUUID := fStr[0]
+		appLogURL := fmt.Sprintf("apps/instanceid/%s/newlogs", appUUID)
```
Do we need the same logic as below? https://github.com/lf-edge/eve/blob/71d286e6fedbf41b795fbc463473ccb7d14493a6/pkg/pillar/cmd/logmanager/logmanager.go#L1029-L1036

naiming-zededa replied:
@giggsoff if you were referring to the v1 vs v2 logic, this newlog scheme only supports V2API

giggsoff replied:
Please take a look https://github.com/lf-edge/adam/pull/50#discussion_r532398858

giggsoff commented:
@naiming-zededa I can see in Adam, that there is no devID inside Comment of gzip:
```
2020/11/30 08:50:00 {"image":"IMGA","eveVersion":"0.0.0-naiming-newlog-infra-f5db03d6-kvm-amd64"}
2020/11/30 08:50:00 Failed to write logbundle message: unable to retrieve valid device UUID from message as : uuid: incorrect UUID length: 
```
Is it expected behavior? Should I calculate devID from the certificate of request?

naiming-zededa commented:
> @naiming-zededa I can see in Adam, that there is no devID inside Comment of gzip:
> 
> ```
> 2020/11/30 08:50:00 {"image":"IMGA","eveVersion":"0.0.0-naiming-newlog-infra-f5db03d6-kvm-amd64"}
> 2020/11/30 08:50:00 Failed to write logbundle message: unable to retrieve valid device UUID from message as : uuid: incorrect UUID length: 
> ```
> 
> Is it expected behavior? Should I calculate devID from the certificate of request?

@giggsoff No, this is not expected behavior, it should be there.

naiming-zededa commented:
> > > all those i think handled from containerd which takes the stdout/stderr and converting them into memlogd input. this is the same for current logging pipeline. The syslog UDP port is between rsyslogd -> logmanager, as a way for logmanager to be the 'server' of this rsyslog.
> > 
> > 
> > I don't think so. @gkodali-zededa don't we depend on rsyslogd listening on the syslog UDP port?
> > @naiming-zededa you should see some dnsmasq logs in the backend. Do you see those with newlog?
> 
> let me recheck on this., the rsyslogd did have a "/dev/log" in configure, I didn't see things from that channel and i removed that. I'll put it back to see then.

@eriknordmark good catch on this. the system's syslog socket is defined as "/dev/log", somehow I had the code but didn't see anything came out, thus I removed this code thinking everything is now into memlogd. added in this commit. I do see 'dnsmasq' and 'dncpcd' log outputs in this socket.

rvs commented:
> @eriknordmark good catch on this. the system's syslog socket is defined as "/dev/log", somehow I had the code but didn't see anything came out, thus I removed this code thinking everything is now into memlogd. added in this commit. I do see 'dnsmasq' and 'dncpcd' log outputs in this socket.

Well, actually.... ;-) @naiming-zededa could we use this as an opportunity to streamline that part of logging as well? I mean there's no reason anything should be using syslog-style logging (also `/dev/log`) since it creates all these additional dataflows in the system that we would then have to control.

I mean -- reading out of that socket is not a big deal (and the right thing to do for now) but should we also commit to a bigger story around streamlining then?

naiming-zededa commented:
> > @eriknordmark good catch on this. the system's syslog socket is defined as "/dev/log", somehow I had the code but didn't see anything came out, thus I removed this code thinking everything is now into memlogd. added in this commit. I do see 'dnsmasq' and 'dncpcd' log outputs in this socket.
> 
> Well, actually.... ;-) @naiming-zededa could we use this as an opportunity to streamline that part of logging as well? I mean there's no reason anything should be using syslog-style logging (also `/dev/log`) since it creates all these additional dataflows in the system that we would then have to control.
> 
> I mean -- reading out of that socket is not a big deal (and the right thing to do for now) but should we also commit to a bigger story around streamlining then?

Roman maybe this steamlining can happen together with your kernel based logging system :-)

rvs commented:
> Roman maybe this steamlining can happen together with your kernel based logging system :-)

I'd be +1 to that -- for sure! Just curious as to what @naiming-zededa thinks of doing the streamlining vs. leaving everything as is.

naiming-zededa commented:
> > Roman maybe this steamlining can happen together with your kernel based logging system :-)
> 
> I'd be +1 to that -- for sure! Just curious as to what @naiming-zededa thinks of doing the streamlining vs. leaving everything as is.

I suspect those daemons don't offer command line options to switch the logging to different places, so we need to have someone reading this socket from, then use newlogd to read this socket is also fine.

rvs commented:
> I suspect those daemons don't offer command line options to switch the logging to different places

actually IIRC at least `dnsmasq` and `dncpcd` do have that option out of the box. The bigger question is that IF (to @eriknordmark point) we ever want this in-kernel memlogd replacement to happen (hopefully this year) we will need to route everything into that location anyway.

Choices are: handle it now (not as part of this PR of course) or handle it later (when we do in-kernel memlogd equivalent).

eriknordmark commented:
> @eriknordmark -- at this point we don't have any freestanding daemons left -- everything gets plumbed through individual containers->containerd->memlogd

@rvs are you sure dnsmasq, radvd and dhcpcd work that way?
It should be easy to check by booting a device with this PR and verifying that dnsmasq and dhcpcd log entries appear at the controller.

naiming-zededa commented:
@giggsoff just committed, the config item is "newlog.allow.fastupload" boolean, if enabled, the log uploading should be similar the the original logging.

rvs commented:
> @giggsoff just committed, the config item is "newlog.allow.fastupload" boolean, if enabled, the log uploading should be similar the the original logging.

Awesome @naiming-zededa !

@giggsoff -- when do you think you can try this with Adam/Eden and see if this passes?

giggsoff commented on docs/CONFIG-PROPERTIES.md:
```diff
@@ -33,6 +33,7 @@
 | timer.appcontainer.stats.interval | integer in seconds | 300 | collect application container stats |
 | timer.vault.ready.cutoff | integer in seconds | 300 | reboot after inaccessible vault |
 | maintenance.mode | "enabled" or "disabled" | "none" | don't run applications etc |
+| newlog.allow.fastupload | "enabled" or "disabled" | disabled | allow faster upload gzip logfiles to controller |
```
As I can see [here](https://github.com/naiming-zededa/eve/blob/7bd6221e180871c2515df73be2836ef65a454f8f/pkg/pillar/types/global.go#L729), newlog.allow.fastupload defined as BoolItem, not TriStateItem. Seems, that we need to change the line.

giggsoff replied:
With `newlog.allow.fastupload: true` it works well as a point of Eden.
So, we need to accept https://github.com/lf-edge/adam/pull/50 as a next step.

naiming-zededa replied:
> As I can see [here](https://github.com/naiming-zededa/eve/blob/7bd6221e180871c2515df73be2836ef65a454f8f/pkg/pillar/types/global.go#L729), newlog.allow.fastupload defined as BoolItem, not TriStateItem. Seems, that we need to change the line.

@giggsoff fixed, thanks.

rvs commented:
Hey @giggsoff with yours and @deitch PRs on the Adam side approved, any chance we can do our weekly release of Eden on Fri (tomorrow) that ties all of this together?

That way, when it comes to our Fri here I can hook it up to EVE's PR checks and make sure that this PR passes Eden and then can (finally!) be merged.

giggsoff commented:
Actually, it is ready and I wait for the new tag of Adam: https://github.com/itmo-eve/eden/tree/newlogs (and some refactoring maybe).

giggsoff commented:
Sad, but true: Eden [indicates](https://github.com/lf-edge/eden/pull/409) problems with reboot. 
I test it with my own and what I see: after onboarding of my device (it works well), I connect to it via console and fire `reboot now`. And I see, that zedbox crashes after restarting of EVE:
```
55fa13b6-0e38-476a-8582-918f259a9442:~# cat /var/persist/log/reboot-stack.log 
goroutine 1 [running]:
github.com/lf-edge/eve/pkg/pillar/agentlog.getStacks(0x0, 0x0, 0x0)
        /pillar/agentlog/agentlog.go:454 +0x78
github.com/lf-edge/eve/pkg/pillar/agentlog.printStack(0xc000390020, 0x7ffffdb38f53, 0xa, 0x2bcf)
        /pillar/agentlog/agentlog.go:212 +0x38
github.com/lf-edge/eve/pkg/pillar/agentlog.initImpl.func1()
        /pillar/agentlog/agentlog.go:636 +0x45
github.com/sirupsen/logrus.runHandler(0xc00044a120)
        /pillar/vendor/github.com/sirupsen/logrus/alt_exit.go:39 +0x49
github.com/sirupsen/logrus.runHandlers()
        /pillar/vendor/github.com/sirupsen/logrus/alt_exit.go:44 +0x4b
github.com/sirupsen/logrus.(*Logger).Exit(0xc0003e2000, 0x1)
        /pillar/vendor/github.com/sirupsen/logrus/logger.go:337 +0x25
github.com/sirupsen/logrus.(*Entry).Fatalf(0xc00026c8c0, 0xc000024160, 0xa4, 0x0, 0x0, 0x0)
        /pillar/vendor/github.com/sirupsen/logrus/entry.go:363 +0x7f
github.com/lf-edge/eve/pkg/pillar/base.(*LogObject).Fatalf(0xc000390020, 0xc000024160, 0xa4, 0x0, 0x0, 0x0)
        /pillar/base/log.go:115 +0x85
main.runService(0x7ffffdb38f53, 0xa, 0x19c6688, 0xc0003e2000, 0xc000390000, 0x1589501)
        /pillar/zedbox/zedbox.go:158 +0x47b
main.main()
        /pillar/zedbox/zedbox.go:125 +0x194
```
```
55fa13b6-0e38-476a-8582-918f259a9442:~# cat /var/persist/log/reboot-reason.log 
Reboot from agent zedmanager[11215] in partition IMGA EVE version 0.0.0-naiming-newlog-infra-54d628b2-kvm-amd64 at 2020-12-04T11:35:39.756973879Z: fatal: agent zedmanager[11215]: publish(/var/run/zedbox/ServiceInitStatus.sock): exception while dialing socket. dial unixpacket /var/run/zedbox/ServiceInitStatus.sock: connect: connection refused
```
I will take a look.

giggsoff commented:
@naiming-zededa seems, that we try to call Stop on nil
```
{"severity":"info","source":"pillar","content":"panic: runtime error: invalid memory address or nil pointer dereference\n","msgid":917,"timestamp":{"seconds":1607083500}}
{"severity":"info","source":"pillar","content":"[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x130ff3f]\n","msgid":918,"timestamp":{"seconds":1607083500}}
{"severity":"info","source":"pillar","content":"\n","msgid":919,"timestamp":{"seconds":1607083500}}
{"severity":"info","source":"pillar","content":"goroutine 1120 [running]:\n","msgid":920,"timestamp":{"seconds":1607083500}}
{"severity":"info","source":"pillar","content":"time.(*Timer).Stop(...)\n","msgid":921,"timestamp":{"seconds":1607083500}}
{"severity":"info","source":"pillar","content":"\t/usr/local/go/src/time/sleep.go:76\n","msgid":922,"timestamp":{"seconds":1607083500}}
{"severity":"info","source":"pillar","content":"github.com/lf-edge/eve/pkg/pillar/cmd/loguploader.handleGlobalConfigImp(0x1632900, 0xc0009b0c40, 0xc000c5b4ed, 0x6, 0x178d320, 0xc0003d1490)\n","msgid":923,"timestamp":{"seconds":1607083500}}
{"severity":"info","source":"pillar","content":"\t/pillar/cmd/loguploader/loguploader.go:434 +0x23f\n","msgid":924,"timestamp":{"seconds":1607083500}}
{"severity":"info","source":"pillar","content":"github.com/lf-edge/eve/pkg/pillar/cmd/loguploader.handleGlobalConfigCreate(0x1632900, 0xc0009b0c40, 0xc000c5b4ed, 0x6, 0x178d320, 0xc0003d1490)\n","msgid":925,"timestamp":{"seconds":1607083500}}
{"severity":"info","source":"pillar","content":"\t/pillar/cmd/loguploader/loguploader.go:410 +0x5d\n","msgid":926,"timestamp":{"seconds":1607083500}}
{"severity":"info","source":"pillar","content":"github.com/lf-edge/eve/pkg/pillar/pubsub.handleModify(0x187d360, 0xc000678000, 0xc000c5b4ed, 0x6, 0xc0004cb800, 0x156d, 0x176d)\n","msgid":927,"timestamp":{"seconds":1607083500}}
{"severity":"info","source":"pillar","content":"\t/pillar/pubsub/subscribe.go:244 +0x67d\n","msgid":928,"timestamp":{"seconds":1607083500}}
{"severity":"info","source":"pillar","content":"github.com/lf-edge/eve/pkg/pillar/pubsub.(*SubscriptionImpl).populate(0xc000678000)\n","msgid":929,"timestamp":{"seconds":1607083500}}
{"severity":"info","source":"pillar","content":"\t/pillar/pubsub/subscribe.go:96 +0x2a5\n","msgid":930,"timestamp":{"seconds":1607083500}}
{"severity":"info","source":"pillar","content":"github.com/lf-edge/eve/pkg/pillar/pubsub.(*SubscriptionImpl).Activate(0xc000678000, 0x19c7d70, 0x19c7d80)\n","msgid":931,"timestamp":{"seconds":1607083500}}
{"severity":"info","source":"pillar","content":"\t/pillar/pubsub/subscribe.go:53 +0x6a\n","msgid":932,"timestamp":{"seconds":1607083500}}
{"severity":"info","source":"pillar","content":"github.com/lf-edge/eve/pkg/pillar/cmd/loguploader.Run(0xc00068cf60, 0xc0001d01c0, 0xc000480b80, 0x0)\n","msgid":933,"timestamp":{"seconds":1607083500}}
{"severity":"info","source":"pillar","content":"\t/pillar/cmd/loguploader/loguploader.go:148 +0x6f5\n","msgid":934,"timestamp":{"seconds":1607083500}}
{"severity":"info","source":"pillar","content":"main.startAgentAndDone(0x19c7d50, 0xc00007b900, 0xc000302ea0, 0xb, 0xc00068cf60, 0xc0001d01c0, 0xc000480b80)\n","msgid":935,"timestamp":{"seconds":1607083500}}
{"severity":"info","source":"pillar","content":"\t/pillar/zedbox/zedbox.go:234 +0x62\n","msgid":936,"timestamp":{"seconds":1607083500}}
{"severity":"info","source":"pillar","content":"created by main.handleService\n","msgid":937,"timestamp":{"seconds":1607083500}}
{"severity":"info","source":"pillar","content":"\t/pillar/zedbox/zedbox.go:224 +0x38a\n","msgid":938,"timestamp":{"seconds":1607083500}}

```
So, please do something like:
```
diff --git a/pkg/pillar/cmd/loguploader/loguploader.go b/pkg/pillar/cmd/loguploader/loguploader.go
index 5c98188f3..848fc39cd 100644
--- a/pkg/pillar/cmd/loguploader/loguploader.go
+++ b/pkg/pillar/cmd/loguploader/loguploader.go
@@ -431,7 +431,9 @@ func handleGlobalConfigImp(ctxArg interface{}, key string, statusArg interface{}
                enabled := gcp.GlobalValueBool(types.AllowLogFastupload)
                if enabled != ctx.enableFastUpload {
                        // reset the schedule for next 30 minutes
-                       ctx.scheduleTimer.Stop()
+                       if ctx.scheduleTimer != nil {
+                               ctx.scheduleTimer.Stop()
+                       }
                        ctx.scheduleTimer = time.NewTimer(1 * time.Second)
                }
                ctx.enableFastUpload = enabled

```

naiming-zededa commented:
@giggsoff good catch, will fix.

naiming-zededa commented:
@giggsoff committed the fix

giggsoff commented:
It works now, thank you.

eriknordmark commented on pkg/newlog/cmd/newlogd.go:
```diff
@@ -0,0 +1,1423 @@
+				Function:  entry.function,
+				Timestamp: timeS,
+			}
+			mapJentry, _ := json.Marshal(mapLog)
```
Will yetus stop complaining if you json.Marshal(&mapLog)?
That should produce the same json output but pass by reference.

naiming-zededa replied:
i'll try. is there way to run yetus locally I'm wondering.

naiming-zededa commented:
added the search for panic stacks for files /persist/reboot-reason and /persist/reboot-stack. although newlog files will get into controller side, for development purpose, it's much easy to ssh into the device to see the stacks. this commit add a new directory /persist/newlog/panicStacks, to keep the last 100 pillar stack files with meta-data, something like:
b9da544a-47c3-4b7c-8c6c-81eb658aa708:/persist# cat newlog/panicStacks/pillar-panic-stack.1607136915
{"devID":"b9da544a-47c3-4b7c-8c6c-81eb658aa708","image":"IMGA","eveVersion":"0.0.0-naiming-newlog-infra-40835cb6-dirty-2020-12-05.02.04-xen-amd64"}
2020-12-05T02:55:15Z,pillar;panic: runtime error: invalid memory address or nil pointer dereference
2020-12-05T02:55:15Z,pillar;[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x131ec47]
2020-12-05T02:55:15Z,pillar;
2020-12-05T02:55:15Z,pillar;goroutine 183 [running]:
2020-12-05T02:55:15Z,pillar;time.(*Timer).Stop(...)
2020-12-05T02:55:15Z,pillar;    /usr/local/go/src/time/sleep.go:76
2020-12-05T02:55:15Z,pillar;github.com/lf-edge/eve/pkg/pillar/cmd/nim.Run(0xc000628420, 0xc00046b030, 0xc000656d80, 0x0)
2020-12-05T02:55:15Z,pillar;    /pillar/cmd/nim/nim.go:769 +0x4be7
2020-12-05T02:55:15Z,pillar;main.startAgentAndDone(0x19c7dc0, 0xc00006b800, 0xc000654b58, 0x3, 0xc000628420, 0xc00046b030, 0xc000656d80)
2020-12-05T02:55:15Z,pillar;    /pillar/zedbox/zedbox.go:234 +0x62
2020-12-05T02:55:15Z,pillar;created by main.handleService
2020-12-05T02:55:15Z,pillar;    /pillar/zedbox/zedbox.go:224 +0x38a
b9da544a-47c3-4b7c-8c6c-81eb658aa708:/persist#

eriknordmark reviewed (approved):
LGTM. Let me approve this to have Eden tests run.

giggsoff commented:
It is the same error (Eden/Adam side). Actually, we are waiting for new image of Adam in docker hub to modify Eden. Today, we solved some inconsistency inside Adam (https://github.com/lf-edge/adam/pull/55), but, I thing, only @rvs can publish it.

naiming-zededa commented:
> It is the same error (Eden/Adam side). Actually, we are waiting for new image of Adam in docker hub to modify Eden. Today, we solved some inconsistency inside Adam ([lf-edge/adam#55](https://github.com/lf-edge/adam/pull/55)), but, I thing, only @rvs can publish it.

got it. thanks.
